File: python/src/minscheme/builtins.py
from typing import Callable, Dict, List

from minscheme import data


def _builtin_print(args: List[data.Value]) -> data.Value:
    assert len(args) == 1
    value = args[0]
    print(str(value))
    return data.UNDEF


def _builtin_add(args: List[data.Value]) -> data.Value:
    assert args
    result = 0
    for value in args:
        assert isinstance(value, data.Int)
        result += value.raw_value
    return data.Int(result)


def _builtin_sub(args: List[data.Value]) -> data.Value:
    assert args
    first_value = args[0]
    assert isinstance(first_value, data.Int), 'not integer'
    result = first_value.raw_value
    for value in args[1:]:
        assert isinstance(value, data.Int), 'not integer'
        result -= value.raw_value
    return data.Int(result)


def _builtin_mul(args: List[data.Value]) -> data.Value:
    assert args
    result = 1
    for value in args:
        assert isinstance(value, data.Int), 'not integer'
        result *= value.raw_value
    return data.Int(result)


def _builtin_div(args: List[data.Value]) -> data.Value:
    assert args
    first_value = args[0]
    assert isinstance(first_value, data.Int), 'not integer'
    result = first_value.raw_value
    for value in args[1:]:
        assert isinstance(value, data.Int), 'not integer'
        result //= value.raw_value
    return data.Int(result)


def _builtin_eq(args: List[data.Value]) -> data.Value:
    assert len(args) == 2
    lhs, rhs = args
    assert isinstance(lhs, data.Int), 'not integer'
    assert isinstance(rhs, data.Int), 'not integer'
    return data.Bool(lhs.raw_value == rhs.raw_value)


def _builtin_lt(args: List[data.Value]) -> data.Value:
    assert len(args) == 2
    lhs, rhs = args
    assert isinstance(lhs, data.Int), 'not integer'
    assert isinstance(rhs, data.Int), 'not integer'
    return data.Bool(lhs.raw_value < rhs.raw_value)


def _builtin_lte(args: List[data.Value]) -> data.Value:
    assert len(args) == 2
    lhs, rhs = args
    assert isinstance(lhs, data.Int), 'not integer'
    assert isinstance(rhs, data.Int), 'not integer'
    return data.Bool(lhs.raw_value <= rhs.raw_value)


def _builtin_gt(args: List[data.Value]) -> data.Value:
    assert len(args) == 2
    lhs, rhs = args
    assert isinstance(lhs, data.Int), 'not integer'
    assert isinstance(rhs, data.Int), 'not integer'
    return data.Bool(lhs.raw_value > rhs.raw_value)


def _builtin_gte(args: List[data.Value]) -> data.Value:
    assert len(args) == 2
    lhs, rhs = args
    assert isinstance(lhs, data.Int), 'not integer'
    assert isinstance(rhs, data.Int), 'not integer'
    return data.Bool(lhs.raw_value >= rhs.raw_value)


def _builtin_and(args: List[data.Value]) -> data.Value:
    result = True
    for value in args:
        result = result and value != data.FALSE
    return data.Bool(result)


def _builtin_or(args: List[data.Value]) -> data.Value:
    result = False
    for value in args:
        result = result or value != data.FALSE
    return data.Bool(result)


def _builtin_not(args: List[data.Value]) -> data.Value:
    assert len(args) == 1
    value = args[0]
    return data.Bool(not (value != data.FALSE))


def _builtin_eq_check(args: List[data.Value]) -> data.Value:
    assert len(args) == 2
    lhs, rhs = args
    return data.Bool(lhs is rhs)


def _builtin_cons(args: List[data.Value]) -> data.Value:
    assert len(args) == 2
    return data.Pair(args[0], args[1])


def _builtin_car(args: List[data.Value]) -> data.Value:
    assert len(args) == 1
    value = args[0]
    assert isinstance(value, data.Pair), 'not pair'
    return value.car


def _builtin_cdr(args: List[data.Value]) -> data.Value:
    assert len(args) == 1
    value = args[0]
    assert isinstance(value, data.Pair), 'not pair'
    return value.cdr


ALL_MAP: Dict[str, Callable[[List[data.Value]], data.Value]] = {
    'print': _builtin_print,
    '+': _builtin_add,
    '-': _builtin_sub,
    '*': _builtin_mul,
    '/': _builtin_div,
    '=': _builtin_eq,
    '<': _builtin_lt,
    '<=': _builtin_lte,
    '>': _builtin_gt,
    '>=': _builtin_gte,
    'and': _builtin_and,
    'or': _builtin_or,
    'not': _builtin_not,
    'eq?': _builtin_eq_check,
    'cons': _builtin_cons,
    'car': _builtin_car,
    'cdr': _builtin_cdr,
}

File: haskell/app/MinScheme/Builtins.hs
module MinScheme.Builtins(
  installBuiltins,
) where

import Data.IORef
import MinScheme.Data
import MinScheme.Environment

valueToInt :: Value -> Int
valueToInt (Integer a) = a
valueToInt _ = error "not an integer"

builtinPrint :: [Value] -> IO Value
builtinPrint [arg] = do
  valueToString arg >>= putStrLn
  return Undef
builtinPrint _ = error "print"

builtinAnd :: [Value] -> Value
builtinAnd args = Boolean $ all valueToBool args

builtinOr :: [Value] -> Value
builtinOr args = Boolean $ any valueToBool args

builtinNot :: [Value] -> Value
builtinNot [arg] = Boolean $ not $ valueToBool arg
builtinNot _ = error "not"

builtinEq :: [Value] -> Value
builtinEq [Integer a, Integer b] = Boolean $ a == b
builtinEq _ = error "=="

builtinLt :: [Value] -> Value
builtinLt [Integer a, Integer b] = Boolean $ a < b
builtinLt _ = error "<"

builtinLte :: [Value] -> Value
builtinLte [Integer a, Integer b] = Boolean $ a <= b
builtinLte _ = error "<="

builtinGt :: [Value] -> Value
builtinGt [Integer a, Integer b] = Boolean $ a > b
builtinGt _ = error ">"

builtinGte :: [Value] -> Value
builtinGte [Integer a, Integer b] = Boolean $ a >= b
builtinGte _ = error ">="

builtinAdd :: [Value] -> Value
builtinAdd values = Integer $ sum $ map valueToInt values

builtinSub :: [Value] -> Value
builtinSub (firstValue:restValues) = Integer $ (valueToInt firstValue -) $ sum $ map valueToInt restValues
builtinSub [] = error "-"

builtinMul :: [Value] -> Value
builtinMul values = Integer $ product $ map valueToInt values

builtinDiv :: [Value] -> Value
builtinDiv (firstValue:restValues) = Integer $ (valueToInt firstValue `div`) $ product $ map valueToInt restValues
builtinDiv [] = error "/"

builtinEqCheck :: [Value] -> Value
builtinEqCheck [Boolean a, Boolean b] | a == b = Boolean True
builtinEqCheck [Integer a, Integer b] | a == b = Boolean True
builtinEqCheck [Symbol a, Symbol b] | a == b = Boolean True
builtinEqCheck [Null, Null] = Boolean True
builtinEqCheck [_, _] = Boolean False
builtinEqCheck _ = error "eq?"

builtinCons :: [Value] -> IO Value
builtinCons [car, cdr] = do
  car' <- newIORef car
  cdr' <- newIORef cdr
  return $ Pair car' cdr'
builtinCons _ = error "cons"

builtinCar :: [Value] -> IO Value
builtinCar [Pair car _] = readIORef car
builtinCar _ = error "car"

builtinCdr :: [Value] -> IO Value
builtinCdr [Pair _ cdr] = readIORef cdr
builtinCdr _ = error "cdr"

installBuiltins :: Env -> IO ()
installBuiltins env = do
  installIO "print" builtinPrint
  installPure "and" builtinAnd
  installPure "or" builtinOr
  installPure "not" builtinNot
  installPure "=" builtinEq
  installPure "<" builtinLt
  installPure "<=" builtinLte
  installPure ">" builtinGt
  installPure ">=" builtinGte
  installPure "+" builtinAdd
  installPure "-" builtinSub
  installPure "*" builtinMul
  installPure "/" builtinDiv
  installPure "eq?" builtinEqCheck
  installIO "cons" builtinCons
  installIO "car" builtinCar
  installIO "cdr" builtinCdr
  where
    installIO name func = do
      var <- ensure env name
      writeIORef var $ Function name func
    installPure name func = installIO name $ return . func
